framework/src/modules/chain/bft/bft.js

```diff
@@ -84,12 +85,62 @@ class BFT extends EventEmitter {
 			this.constants.startingHeight,
 		);
 
-		await this.loadBlocks({
+		await this.loadBlocksFromStorage({
 			fromHeight: loadFromHeight,
 			tillHeight: lastBlockHeight,
 		});
 	}
 
+	/**
+	 * When blocks deleted send those to BFT to update BFT state
+	 *
+	 * @param {Array.<Object>} blocks - List of all blocks
+	 * @return {Promise<void>}
+	 */
+	async onDeleteBlocks(blocks) {
+		assert(blocks, 'Must provide blocks which are deleted');
+		assert(Array.isArray(blocks), 'Must provide list of blocks');
+
+		// We need only height to delete the blocks
+		// But for future extension we accept full blocks in BFT
+		// We may need to utilize some other attributes for internal processing
+		const blockHeights = blocks.map(({ height }) => height);
+
+		assert(
+			!blockHeights.some(h => h <= this.finalityManager.finalizedHeight),
+			'Can not delete block below or same as finalized height',
+		);
+
+		const removeFromHeight = Math.min(...blockHeights);
+
+		this.finalityManager.removeBlockHeaders({
+			aboveHeight: removeFromHeight - 1,
+		});
+
+		// Make sure there are 2 rounds of block headers available
+		if (
+			this.finalityManager.maxHeight - this.finalityManager.minHeight <
+			this.constants.activeDelegates * 2
+		) {
+			const tillHeight = this.finalityManager.minHeight - 1;
+			const fromHeight =
+				this.finalityManager.maxHeight - this.constants.activeDelegates * 2;
+			await this.loadBlocksFromStorage({ fromHeight, tillHeight });
+		}
+	}
+
+	/**
+	 * Load new block to BFT
+	 *
+	 * @param {Object} block - The block which is forged
+	 * @return {Promise<void>}
+	 */
+	async onNewBlock(block) {
```

Comment:
Here as well.

Reply:
@SargeKhan Its not an event. Its a function that you can invoke or bound to an event.

Reply:
We have the convention of using `on` for functions which are triggered via events in the chain module. This function should be executed as part of block processing. Perhaps we can update the name to `addNewBlockHeader` in order to add further clarity?

Reply:
I don't see any such conventions around project in current state. But idea was similar that `onNewBlock` should be bound to event outside BFT.